storage/pebble/txn.go

```
@@ -0,0 +1,127 @@
+	db    *PebbleDB
+}
+
+func newPebbleTxn(batch *pebble.Batch, db *PebbleDB) basedb.Txn {
+	return &pebbleTxn{
+		batch: batch,
+		db:    db,
+	}
+}
+
+func (t *pebbleTxn) Commit() error {
+	return t.batch.Commit(pebble.Sync)
+}
+
+func (t *pebbleTxn) Discard() {
+	_ = t.batch.Close()
+}
+
+func (t *pebbleTxn) Set(prefix []byte, key []byte, value []byte) error {
+	return t.batch.Set(append(prefix, key...), value, nil)
+}
+
+func (t *pebbleTxn) SetMany(prefix []byte, n int, next func(int) (basedb.Obj, error)) error {
+	for i := 0; i < n; i++ {
+		item, err := next(i)
+		if err != nil {
+			return err
+		}
+		if err := t.batch.Set(append(prefix, item.Key...), item.Value, nil); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (t *pebbleTxn) Get(prefix []byte, key []byte) (basedb.Obj, bool, error) {
+	value, closer, err := t.batch.Get(append(prefix, key...))
+	if err != nil {
+		if errors.Is(err, pebble.ErrNotFound) {
+			return basedb.Obj{}, false, nil
+		}
+		return basedb.Obj{}, true, err
+	}
+	defer func() { _ = closer.Close() }()
+	valCopy := make([]byte, len(value))
+	copy(valCopy, value)
+	return basedb.Obj{
+		Key:   key,
+		Value: valCopy,
+	}, true, nil
+}
+
+func (t *pebbleTxn) GetMany(prefix []byte, keys [][]byte, iterator func(basedb.Obj) error) error {
+	for _, key := range keys {
+		fullKey := append(prefix, key...)
+		value, closer, err := t.batch.Get(fullKey)
+		if err != nil {
+			if errors.Is(err, pebble.ErrNotFound) {
+				continue
+			}
+			return err
+		}
+
+		valCopy := make([]byte, len(value))
+		copy(valCopy, value)
+		if err := closer.Close(); err != nil {
+			return err
+		}
+		obj := basedb.Obj{
+			Key:   key,
+			Value: valCopy,
+		}
+		if err := iterator(obj); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (t *pebbleTxn) GetAll(prefix []byte, fn func(int, basedb.Obj) error) error {
+	iter, err := makePrefixIter(t.batch, prefix)
+	if err != nil {
+		return err
+	}
+
+	defer func() { _ = iter.Close() }()
```

i think we should return an error if this failed to close (it might do more than just closed in some cases)

---

the errors returned from the `.Close()` method are the same as the errors returned from the `iter.Error()` at the end of the function, so we're not ignoring them.

---

the only reason I defer the call to `Close()` is because - if a panic happens in the callback function - the defer statement will be executed regardless, but if we return `Close()` instead of `iter.Error()` that would lead to a leak.

---

We could keep using `defer` if we want and handle error from `Close` in there ... 

but I wouldn't do that - it's a bit bulky and IMO unnecessary since `panic` in callback function would mean we have a bug in the code that needs to be addressed, I imagine not calling `Close` in this case is of smaller importance

---

I think the conclusion of this [article](https://dev.ribic.ba/defer-mutex-unlocks/) got stuck with me too much and that's why I went by the route of deferring `Close` statements here.

---

Yeah `defer` is a very nice thing to have/use - I don't mind it, 

although here it seems to me `panic` is unlikely to happen without us noticing it during testing